[Navigation API] NavigateEvent for location.href needs to run before next microtask.

basuke opened the pull request:
#### db6e4c26e758915244eae2cf0f8d03ca66a59ec6
<pre>
[Navigation API] NavigateEvent for location.href needs to run before next microtask.
<a href="https://bugs.webkit.org/show_bug.cgi?id=296945">https://bugs.webkit.org/show_bug.cgi?id=296945</a>
<a href="https://rdar.apple.com/157307363">rdar://157307363</a>

Reviewed by Alex Christensen.

When location.href or similar APIs trigger a navigation, the navigate event should fire
synchronously before any microtasks run. This ensures proper event ordering and allows
navigate event handlers to prevent the navigation if needed. The fix adds synchronous
NavigateEvent dispatch in scheduleLocationChange().

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-canceled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-double-intercept_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-double-intercept_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reentrant_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reentrant_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reject_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reject_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-finished-mark-as-handled-expected.txt:
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::skipNavigateEvent const):
(WebCore::FrameLoadRequest::setSkipNavigateEvent):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadPostRequest):
(WebCore::FrameLoader::dispatchNavigateEvent):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledLocationChange::ScheduledLocationChange):
(WebCore::NavigationScheduler::scheduleLocationChange):

Canonical link: <a href="https://commits.webkit.org/298445@main">https://commits.webkit.org/298445@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b4755d25e914aaa47282ff5d4ef8a40f8c2daa98

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115494 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35178 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/25687 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121570 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66056 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c6fbac06-cdd9-466b-be17-e966673f0f0a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117383 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/35845 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/43754 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/87747 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b41bca31-4402-49b2-9477-3e67359ca7bf) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118442 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/103662 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68140 "Passed tests") | | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/114868 "Build is in progress. Recent messages:") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/27747 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/21782 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65234 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/97980 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/21900 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/124732 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42420 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/31788 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/96522 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/42787 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/99853 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96308 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24505 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41545 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19404 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38312 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42303 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/47865 "Built successfully") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/41800 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45128 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43516 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebCore/loader/NavigationScheduler.cpp:
```diff
@@ -606,11 +609,35 @@ void NavigationScheduler::scheduleLocationChange(Document& initiatingDocument, S
+                    if (RefPtr navigation = window->navigation()) {
+                        auto navigationType = (historyHandling == NavigationHistoryBehavior::Replace) ? NavigationNavigationType::Replace : NavigationNavigationType::Push;
+
+                        if (!navigation->dispatchPushReplaceReloadNavigateEvent(url, navigationType, false, nullptr, nullptr, nullptr))
```
This function call has a bunch of magic values.  It would be good to give at least "false" a name, and use some of the default values from the header:
constexpr bool isSameDocument { false };
dispatchPushReplaceReloadNavigateEvent(url, navigationType, isSameDocument, nullptr);

Also, this is quite deeply nested.  It would be nice if it could be less deeply nested by doing some things like this:
RefPtr window = document->window();
if (RefPtr navigation = window ? window->navigation() : nullptr)

achristensen07 commented on Source/WebCore/loader/NavigationScheduler.cpp:
```diff
@@ -606,11 +609,35 @@ void NavigationScheduler::scheduleLocationChange(Document& initiatingDocument, S
+    if (localFrame && !url.protocolIsJavaScript()) {
+        RefPtr document = localFrame->document();
+        if (document && document->settings().navigationAPIEnabled()) {
+            bool isSameOrigin = document->securityOrigin().isSameOriginAs(securityOrigin);
```
This result having a variable name isn't useful.  It could just be in the if statement above:
if (document && document->settings().navigationAPIEnabled() && document->securityOrigin().isSameOriginAs(securityOrigin)) {

achristensen07 commented on Source/WebCore/loader/NavigationScheduler.cpp:
```diff
@@ -258,6 +260,7 @@ class ScheduledLocationChange : public ScheduledURLNavigation {
 private:
     CompletionHandler<void(bool)> m_completionHandler;
     NavigationHistoryBehavior m_navigationHistoryBehavior;
+    bool m_hasDispatchedNavigateEvent;
```
May as well give this a default value to avoid using uninitialized memory:
bool m_hasDispatchedNavigateEvent { false };

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1632,15 +1634,18 @@ void FrameLoader::loadURL(FrameLoadRequest&& frameLoadRequest, const String& ref
-            return;
+        if (!skipNavigateEvent) {
+            RefPtr sourceElement = event ? dynamicDowncast<Element>(event->target()) : nullptr;
+            if (!dispatchNavigateEvent(newURL, newLoadType, action.downloadAttribute(), historyHandling, true, formState.get(), nullptr, sourceElement.get()))
```
It would be nice to give "true" a name.

achristensen07 commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1481,12 +1481,14 @@ void FrameLoader::loadFrameRequest(FrameLoadRequest&& request, Event* event, Ref
-                RefPtr sourceElement = event ? dynamicDowncast<Element>(event->target()) : nullptr;
-                if (!dispatchNavigateEvent(url, loadType, request.downloadAttribute(), request.navigationHistoryBehavior(), false, formState.get(), stateObject.get(), sourceElement.get()))
-                    return;
+                if (!request.skipNavigateEvent()) {
```
It seems like kind of an anti pattern to require all callers of dispatchNavigateEvent to check this boolean.  It seems like it would make the code a bit cleaner to just pass in a bool indicating whether the navigation event was dispatched when scheduled.  That would also make it so that future callers won't forget it.

achristensen07 commented on Source/WebCore/page/Navigation.h:
```diff
@@ -145,7 +145,14 @@ class Navigation final : public RefCounted<Navigation>, public EventTarget, publ
     DispatchResult dispatchTraversalNavigateEvent(HistoryItem&);
-    bool dispatchPushReplaceReloadNavigateEvent(const URL&, NavigationNavigationType, bool isSameDocument, FormState*, SerializedScriptValue* classicHistoryAPIState = nullptr, Element* sourceElement = nullptr);
+
+    static constexpr bool isSameDocument { true };
```
No, not here.  Remove all of these and just have regular default values.  What I was saying earlier was to replace these at the call sites:
"if (!dispatchNavigateEvent(loadType, request, false, formState.get(), stateObject.get(), event))"
with something like this:
"constexpr bool isSameDocument = false;"
"if (!dispatchNavigateEvent(loadType, request, isSameDocument, formState.get(), stateObject.get(), event))"
That just makes the call sites a little easier to read.  Replacing bool isSameDocument with a 2-state enum would achieve the same thing.  But naming the default arguments in the header doesn't give more information to people reading the code.

basuke committed:
[Navigation API] NavigateEvent for location.href needs to run before next microtask.
https://bugs.webkit.org/show_bug.cgi?id=296945
rdar://157307363

Reviewed by Alex Christensen.

When location.href or similar APIs trigger a navigation, the navigate event should fire
synchronously before any microtasks run. This ensures proper event ordering and allows
navigate event handlers to prevent the navigation if needed. The fix adds synchronous
NavigateEvent dispatch in scheduleLocationChange().

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-canceled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-double-intercept_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-double-intercept_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reentrant_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reentrant_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reject_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept-reject_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept_currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/location-href-intercept_no-currententrychange-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/ordering-and-transition/transition-finished-mark-as-handled-expected.txt:
* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::skipNavigateEvent const):
(WebCore::FrameLoadRequest::setSkipNavigateEvent):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::loadFrameRequest):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadPostRequest):
(WebCore::FrameLoader::dispatchNavigateEvent):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledLocationChange::ScheduledLocationChange):
(WebCore::NavigationScheduler::scheduleLocationChange):

Canonical link: https://commits.webkit.org/298445@main
